Sources/TSCBasic/misc.swift

```diff
@@ -87,9 +87,20 @@ public func lookupExecutablePath(
 ) -> AbsolutePath? {
 
     // We should have a value to continue.
+#if os(Windows)
+    guard var value = value, !value.isEmpty else {
+        return nil
+    }
+    let isPath = value.contains("\\")
+    if !isPath && !value.contains(".") {
+        value.append(executableFileSuffix)
```

Comment:
Executable do not need to have the `.exe` suffix - case in point the XCTest bundle.

Reply:
Executive don’t need to be `.exe`, but they must have a file extension. If there isn’t one, Windows would assume it to be `.exe` (That’s why there’s a`!value.contains(".")` judgement.

Reply:
I don't think that they need to have an extension.  Use of `CreateProcessW` doesn't care if the image has an extension or not.  I think that the restriction is implemented at the shell level.  Is there documentation that states otherwise?

Reply:
> If the file name does not contain an extension, .exe is appended. 

From https://docs.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-createprocessw

Reply:
> If the file name ends in a period (.) with no extension, or if the file name contains a path, .exe is not appended.

The following sentences are important.  Because we are invoking tools via path, the suffix is not added.

Reply:
Yep, but doesn’t this method means to simulate the path finding process of `CreateProcessW`? That is, when we invoke `lookupExecutablePath("cmd.exe")` we can get `AbsolutePath("C:\\Windows\\System32\\cmd.exe")`, and I think `lookupExecutablePath("cmd")` should have the same output.

This would also help to eliminate platform differences between Windows and POSIX systems.

Reply:
While I agree with the desire to remove the `#if os` checks, I don't agree with the approach.  We already have a way to get the expected extension in a platform agnostic manner, which easily allows us to add the extension without having to conditionalise the code on the OS.  We already know that we expect the files to have the `.exe` suffix.

Alternatively, why not use a number of extensions?  `.cmd` would be a very reasonable wrapper that we might want to use, or perhaps `.ps1`?  But this allows for command injection if there is a site that we missed adding the extension to.

I think I would be against this lookup behaviour change.  Instead, we should audit the sites and use the correct extension.

Reply:
I totally agree that we can use executable file types other than `exe`, but... I don’t think looking for an executable file without extension on Windows is an expected behavior. From the caller’s perspective, one would expect `lookupExecutablePath("clang")` to find `clang.exe` on Windows, instead of finding a `clang` without extension, because both the system API and shell would do so. (And I don’t think we can really invoke Windows scripts from Swift easily, based on the current implementation of `Foundation`?)

Also, `STSC` is a widely-used supporting library, and we need to manually handle the `.exe` suffix platform-specifically nearly everywhere it’s used. That would produce lots of needless codes and may bring more problems. If we adopt the defaulting behavior, hopefully no existing code would break, and anywhere the code breaks there must be a missing suffix in the toolchain. Then we can easily locate and eliminate such problems.

Reply:
I think that if you want to use this to actually find those cases, you should just assert that the path has an extension, which would immediately catch the cases where it does not.  Note that the XCTest bundle does not have an exe suffix, and is currently assumed to be executable.  How are you proposing to handle that?

Reply:
Any executable file type with an extension other than `.exe` will not be affected by this change. That is, you can still use `lookupExecutablePath("TestSomething.xctest")` to find the test bundle, nothing would be different. The only difference is that besides using `lookupExecutablePath("something.exe")`, you can now use `lookupExecutablePath("something")` to retrieve the same result, and that an "executable" without extension would be abandoned.

Reply:
You cannot be certain of the use of the API once it is made available.  There is no way to track down each use and ensure that the use satisfies the constraints.  It is better to force the constraints to avoid having issues like that show up later.

Reply:
On the other side, taking this change is enforcing all executables to have a correct extension on Windows. This is exactly the expected behavior. If we find any executable without extension in the toolchain, there should be a bug.

Reply:
No, there shouldn't be any expectation on the file name.  That's the point.

Yes, the toolchain binaries are suffixed with `.exe`.  But this API does not enforce a search within the toolchain.  If you can force TSC to only search within the toolchain directory and nothing else (which means no cmd, no user generated binaries, etc), adding the suffix would be okay.  But even then, an assertion would be better than the forced suffix.

Reply:
According to Microsoft Docs, an executable without extension cannot be executed with its name through `CreateProcess` either. This is nowhere an expected behavior.

If one persists to use such a file name, he should provide a full path here, like what he’s required in `CreateProcess`.

Reply:
As a workaround, we can add (though I think it unnecessary) an optional Boolean parameter here to indicate whether we should automatically add suffixes to it, and defaulting it to `true`.

Reply:
I think that the default should be `false`.  It should be an explicit opt-in.

Reply:
```
.\a.out --stdlib-unittest-in-process --stdlib-unittest-filter "diff param generic => concrete"
```
Note that it uses the traditional Unix spelling (`a.out`) even though the output is not an "a.out" binary.  Expecting `.exe` suffixes for the tools is reasonable, but embedding that into a library I do not think is the correct approach.  Reserving that knowledge to the site of use is better, as there is nothing which mandates the extension.